File: experiments/09_grid_localization/src/matlab/read_jsonc.m
function config = read_jsonc(filepath)
% READ_JSONC Read JSONC file (JSON with comments) and parse it
% Strips // comments and then uses jsondecode
%
% Input:
%   filepath - Path to .jsonc file
% Output:
%   config - Parsed configuration structure

    % Read entire file
    raw_text = fileread(filepath);
    
    % Split into lines
    lines = splitlines(raw_text);
    
    % Remove // comments and empty lines
    cleaned_lines = {};
    for i = 1:length(lines)
        line = strtrim(lines{i});
        
        % Skip empty lines
        if isempty(line)
            continue;
        end
        
        % Find // comment (but not inside strings)
        % Simple approach: find first // that's not in a string
        comment_pos = strfind(line, '//');
        if ~isempty(comment_pos)
            % Check if it's inside quotes
            in_string = false;
            actual_comment_pos = [];
            for j = 1:length(line)
                if line(j) == '"' && (j == 1 || line(j-1) ~= '\')
                    in_string = ~in_string;
                end
                if j == comment_pos(1) && ~in_string
                    actual_comment_pos = j;
                    break;
                end
            end
            
            if ~isempty(actual_comment_pos)
                % Remove comment part
                line = strtrim(line(1:actual_comment_pos-1));
            end
        end
        
        % Add line if not empty after comment removal
        if ~isempty(line)
            cleaned_lines{end+1} = line; %#ok<AGROW>
        end
    end
    
    % Join back into single string
    json_text = strjoin(cleaned_lines, newline);
    
    % Parse JSON
    config = jsondecode(json_text);
end

File: experiments/09_grid_localization/src/python/utils/read_jsonc.py
"""
Helper to read JSONC files (JSON with comments)
"""
import json
import re


def read_jsonc(filepath):
    """
    Read JSONC file (JSON with comments) and parse it.
    Strips // comments and then uses json.loads
    
    Args:
        filepath: Path to .jsonc file
    
    Returns:
        Parsed configuration dictionary
    """
    with open(filepath, 'r', encoding='utf-8') as f:
        text = f.read()
    
    # Remove single-line comments (// ...)
    # This regex handles comments while preserving // inside strings
    lines = []
    for line in text.split('\n'):
        # Find // that's not inside a string
        in_string = False
        escape_next = False
        comment_pos = None
        
        for i, char in enumerate(line):
            if escape_next:
                escape_next = False
                continue
            
            if char == '\\':
                escape_next = True
                continue
            
            if char == '"':
                in_string = not in_string
            
            if not in_string and i < len(line) - 1:
                if line[i:i+2] == '//':
                    comment_pos = i
                    break
        
        if comment_pos is not None:
            line = line[:comment_pos].rstrip()
        
        lines.append(line)
    
    # Join back and parse
    cleaned_text = '\n'.join(lines)
    return json.loads(cleaned_text)
